client/src/main/java/com/netflix/conductor/client/task/WorkflowTaskCoordinator.java

```diff
@@ -23,426 +23,420 @@
+            LOGGER.debug("Instance is NOT UP in discovery - will not poll");
+            return;
+        }
+
+        if (worker.paused()) {
+            WorkflowTaskMetrics.incrementTaskPausedCount(worker.getTaskDefName());
+            LOGGER.debug("Worker {} has been paused. Not polling anymore!", worker.getClass());
+            return;
+        }
+
+        Task task;
+        try {
+            if (!pollingSemaphore.canPoll()) {
+                return;
+            }
 
-	private List<Worker> workers = new LinkedList<>();
+            String taskType = worker.getTaskDefName();
+            String domain = Optional.ofNullable(PropertyFactory.getString(taskType, DOMAIN, null))
+                .orElse(PropertyFactory.getString(ALL_WORKERS, DOMAIN, null));
 
-	private int sleepWhenRetry;
+            LOGGER.debug("Polling task of type: {} in domain: '{}'", taskType, domain);
+            task = WorkflowTaskMetrics.getPollTimer(taskType)
+                .record(() -> taskClient.pollTask(taskType, worker.getIdentity(), domain));
 
-	private int updateRetryCount;
+            if (Objects.nonNull(task) && StringUtils.isNotBlank(task.getTaskId())) {
+                WorkflowTaskMetrics.incrementTaskPollCount(taskType);
+                LOGGER.debug("Polled task: {} of type: {} in domain: '{}', from worker: {}",
+                    task.getTaskId(), taskType, domain, worker.getIdentity());
 
-	private int workerQueueSize;
+                CompletableFuture<Task> taskCompletableFuture = CompletableFuture.supplyAsync(() ->
+                    processTask(task, worker), executorService);
 
-	private LinkedBlockingQueue<Runnable> workerQueue;
-
-	private int threadCount;
+                taskCompletableFuture.whenComplete(this::finalize);
+            } else {
+                // no task was returned in the poll, release the permit
+                pollingSemaphore.complete();
+            }
+        } catch (Exception e) {
+            // release the permit if exception is thrown during polling, because the thread would not be busy
+            pollingSemaphore.complete();
+            WorkflowTaskMetrics.incrementTaskPollErrorCount(worker.getTaskDefName(), e);
+            LOGGER.error("Error when polling for tasks", e);
+        }
+    }
+
+    private Task processTask(Task task, Worker worker) {
+        LOGGER.debug("Executing task: {} of type: {} in worker: {}", task.getTaskId(), task.getTaskDefName(),
+            worker.getIdentity());
+        try {
+            execute(worker, task);
+        } catch (Throwable t) {
+            task.setStatus(Task.Status.FAILED);
+            TaskResult result = new TaskResult(task);
+            handleException(t, result, worker, task);
+        } finally {
+            pollingSemaphore.complete();
+        }
+        return task;
+    }
+
+    private void execute(Worker worker, Task task) {
+        String taskType = task.getTaskDefName();
+        try {
+            if (!worker.preAck(task)) {
```

Comment:
Should this be still performed ? There should be no `ACK`  semantics since the task is in progress once polled.

Reply:
`TaskPollExecutor` does not use ack anymore